compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.FirSession
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.isSubtypeOfAny
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeTypeParameterInQualifiedAccess
+import org.jetbrains.kotlin.fir.resolve.inference.isKClassType
+import org.jetbrains.kotlin.fir.types.*
+
+// FE1.0 [KClassWithIncorrectTypeArgumentChecker]
+object FirKClassWithIncorrectTypeArgumentChecker : FirFileChecker() {
+    override fun check(declaration: FirFile, context: CheckerContext, reporter: DiagnosticReporter) {
+        for (topLevelDeclaration in declaration.declarations) {
+            if (topLevelDeclaration is FirCallableMemberDeclaration<*>) {
+                checkTopLevelDeclaration(topLevelDeclaration, context, reporter)
+            }
+        }
+    }
+
+    private fun checkTopLevelDeclaration(
+        declaration: FirCallableMemberDeclaration<*>,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        if (declaration.visibility == Visibilities.Local) return
+
+        // prevent duplicate reporting
+        if (declaration is FirPropertyAccessor) return
+
+        val source = declaration.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+
+        val returnType = declaration.returnTypeRef.coneType
+        if (!returnType.isKClassWithBadArgument(context.session)) return
+
+        val typeArgument = (returnType.typeArguments[0] as ConeKotlinTypeProjection).type
+        typeArgument.typeParameterFromError?.let {
+            reporter.reportOn(source, FirErrors.KCLASS_WITH_NULLABLE_TYPE_PARAMETER_IN_SIGNATURE, it.symbol, context)
+        }
+    }
+
+    private fun ConeKotlinType.isKClassWithBadArgument(session: FirSession): Boolean {
+        if (!this.isKClassType()) return false
+        val argumentType = typeArguments.toList().singleOrNull()?.let {
+            when (it) {
+                is ConeStarProjection -> null
+                is ConeKotlinTypeProjection -> it.type
+            }
+        } ?: return false
+        argumentType.typeParameterFromError?.let { typeParameter ->
+            if (typeParameter.bounds.any { it.coneType.isSubtypeOfAny(session) }) {
```

Comment:
Same as comment on other commit: What about the following case:
`fun <T> f3() where T : Any?, T : Comparable<T> = T::class`

I think this should be either `bounds.all` or get the intersection type.

Reply:
Right. Again, simple subtype check is good enough (and match FE 1.0). Added a test as well, thanks!